content/posts/apisix-wasm-support.md

```diff
@@ -0,0 +1,125 @@
+	plugin := ctx.parent
+	for _, hdr := range plugin.Headers {
+		proxywasm.ReplaceHttpResponseHeader(hdr.Name, hdr.Value)
+	}
+
+	return types.ActionContinue
+}
+```
+
+The [SDK exports this function](https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/1b9daaf70730bd6197ca8a34b2a8af713bbce7ad/proxywasm/internal/abi_callback_l7.go#L63) as `proxy_on_response_header` and can now be called from a host environment that implements the Proxy Wasm specification, i.e., Apache APISIX.
+
+When APISIX receives a request and enters the `header_filter` [phase](https://apisix.apache.org/docs/apisix/terminology/plugin/#plugins-execution-lifecycle), it [calls](https://github.com/apache/apisix/blob/1eaad271ecc78ae2001e9da79b347406bd9aa6be/apisix/wasm.lua#L119) the `on_http_response_header` Lua function from the wasm-nginx-module. This function, in turn, [sets the callback function name](https://github.com/api7/wasm-nginx-module/blob/ccc83f7397c711b5f99a55682134a0972a5a6040/src/http/ngx_http_wasm_module.c#L702), `cb_name`, based on the stage of the request lifecycle.
+
+Since the stage is `HTTP_RESPONSE_HEADERS`, the `cb_name` is set to `proxy_on_response_headers`:
+
+```c
+if (type == HTTP_RESPONSE_HEADERS) {
+	cb_name = &proxy_on_response_headers;
+}
+```
+
+As you might have guessed, the `cb_name` corresponds to the function to be called in the Wasm plugin. So the wasm-nginx-module [calls](https://github.com/api7/wasm-nginx-module/blob/ccc83f7397c711b5f99a55682134a0972a5a6040/src/http/ngx_http_wasm_module.c#L710) the `proxy_on_response_headers` function, i.e., the `OnHttpResponseHeaders` function we wrote in Go, through the Wasm VM:
+
+```c
+if (type == HTTP_REQUEST_HEADERS || type == HTTP_RESPONSE_HEADERS) {
+    if (hwp_ctx -> hw_plugin -> abi_version == PROXY_WASM_ABI_VER_010) {
+        rc = ngx_wasm_vm -> call(hwp_ctx -> hw_plugin -> plugin,
+            cb_name,
+            true, NGX_WASM_PARAM_I32_I32, http_ctx -> id, 0);
+    } else {
+        rc = ngx_wasm_vm -> call(hwp_ctx -> hw_plugin -> plugin,
+            cb_name,
+            true, NGX_WASM_PARAM_I32_I32_I32, http_ctx -> id, 0, 1);
+      }
+}
+```
+
+What's a Wasm VM though?
+
+## Wasm Virtual Machines
+
+Wasm binaries are executed in Wasm virtual machines (VMs). They implement the WASI standard, allowing Wasm binaries to be run outside the browser.
+
+The wasm-nginx-module implements two Wasm VMs, [Wasmtime](https://wasmtime.dev/) and [WasmEdge](https://wasmedge.org/), and uses the former as default.
+
+Running Wasm plugins inside a VM comes with beneficial side effects. It offers security and isolation from APISIX, ensuring that any issues with the plugin are self-contained.
+
+The wasm-nginx-module first [loads](https://github.com/api7/wasm-nginx-module/blob/ccc83f7397c711b5f99a55682134a0972a5a6040/lib/resty/proxy-wasm.lua#L81) the `.wasm` binary to the VM and then uses the `call` method, as we saw above. All Wasm plugins are run inside the same Wasm VM.
+
+## Programming Model
```

Comment:
Maybe it is better to put the Programming Model section after the Proxy Wasm Specification. It makes the topic easier to follow in order.

Reply:
That was the plan but I mentioned Wasm VMs and haven't explained what they were. Maybe I can move the Wasm VMs to the beginning and make the move you suggested.

Reply:
Yes, this approach is good!